data manager for bakta tool

pimarin opened the pull request:
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

This is the data manager linked to bakta tool
I failed to understand how to make it, so this is the first try. If you can help me to improve the data_manager
The problem is with bakta database which is a huge database build in one time by the tool.
I discussed with the author but it seem impossible to create a light DB for the test.

pimarin commented:
The bakta_db script allow 3 actions : 

- Download a version from the list
- List all available database from zenodo
- Update database path specified

I created only the download and the update

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,54 @@
+    </outputs>
+    <tests>
+
+        <!-- TEST2 update -->
```
test1? :)

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,54 @@
+            <param name="database_command" value="update"/>
+            <output name="output_file">
+                <assert_contents>
+                    <has_text text="pangoLEARN data release 2021-04-01"/>
```
pangoLearn?

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,54 @@
+            download --output $output_file
+        #else if $database_command == "update"
+        #else if $database_command == "update"
+            update --db $output_file
```
single quote missing

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,54 @@
+    <command detect_errors="aggressive"><![CDATA[
+        python '$__tool_directory__/bakta_build_database.py'
+        #if $database_command == "download"
+            download --output $output_file
```
single quote

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,54 @@
+        python '$__tool_directory__/bakta_build_database.py'
+        #if $database_command == "download"
+            download --output $output_file
+        #else if $database_command == "update"
```
two else-if?

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,54 @@
+            <param name="database_command" type="select" label="Database action" multiple="false" help="Action on bakta database. Install or update database">
+                <option value="download"> Download a prebuild bakta database</option>
+                <option value="custom"> Download a custom database</option>
+                <option value="update"> Update the databases</option>
```
update can be removed isn't it?

bgruening reviewed (commented):
I agree the update functionality does not make much sense in this context. But you still have it in your code as it seems.

The python codes needs some lint-fixes to pass th CI checks.

bgruening commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -8,26 +8,10 @@
-        #else if $database_command == "update"
-            update --db $output_file
-        #end if
+        download --output $output_file
```
singl quotes

pimarin commented:
I'm doing test to understand how to build it but it failed
The data_manager seem to work --> it dowload the `.tar.gz` database but it's the real database (32Go!).
How to deal to make a lightest download ?

pimarin commented:
I asked to the developer if it possible to have a lighted test-database, but it's not possible easily.

pimarin commented:
I failed to make a test:
```
Problems:

    Job in error state.. tool_id: bakta_build_database, exit_code: 0, stderr: .

    Job in error state.. tool_id: bakta_build_database, exit_code: 0, stderr: .

Command Line:

    mkdir out_dir && python '/home/pierre/.planemo/planemo_tmp_fqj_4bp7/data_manager/bakta_build_database.py' download --output 'out_dir/tmp/tmpwg6kthw8/job_working_directory/000/2/outputs/galaxy_dataset_1509a276-3404-4fe2-a150-d697433c9f3a.dat'
```
The command work I have it :
```
fetch DB versions...
	... compatible DB versions: 2
download database: v3.1, 2022-02-03, DOI: 10.5281/zenodo.5961398, URL: https://zenodo.org/record/5961398/files/db.tar.gz...
[?25l[?25h|ââââââââââââââââââââââââââââââââââââââââ| 28593946/28593946 [100%] in 7:08.6 (6
	... done
check MD5 sum...
	...database file OK: 95a140c0c2a40153fdc9794e89860944
extract DB tarball: file=/tmp/tmpwg6kthw8/job_working_directory/000/2/working/out_dir/tmp/tmpwg6kthw8/job_working_directory/000/2/outputs/galaxy_dataset_1509a276-3404-4fe2-a150-d697433c9f3a.dat/db.tar.gz, output=/tmp/tmpwg6kthw8/job_working_directory/000/2/working/out_dir/tmp/tmpwg6kthw8/job_working_directory/000/2/outputs/galaxy_dataset_1509a276-3404-4fe2-a150-d697433c9f3a.dat
successfully downloaded Bakta database!
	version: 3.1
	DOI: 10.5281/zenodo.5961398
	path: /tmp/tmpwg6kthw8/job_working_directory/000/2/working/out_dir/tmp/tmpwg6kthw8/job_working_directory/000/2/outputs/galaxy_dataset_1509a276-3404-4fe2-a150-d697433c9f3a.dat/db
update AMRFinderPlus database...
	... done

Run Bakta using '--db /tmp/tmpwg6kthw8/job_working_directory/000/2/working/out_dir/tmp/tmpwg6kthw8/job_working_directory/000/2/outputs/galaxy_dataset_1509a276-3404-4fe2-a150-d697433c9f3a.dat/db' or set a BAKTA_DB environment variable: 'export BAKTA_DB=/tmp/tmpwg6kthw8/job_working_directory/000/2/working/out_dir/tmp/tmpwg6kthw8/job_working_directory/000/2/outputs/galaxy_dataset_1509a276-3404-4fe2-a150-d697433c9f3a.dat/db'
```
52Go downloaded... but that's all and not test on the json file...

pimarin commented:
Okay...
I understand some problems, and I have some questions...
1. the python script was just the script from the authors to download the database but it is not so simple
    - need to create a real python script available to download the database
    - these explain why I can't test json file
2. A question is, is it possible to have an evolutive tool ?
    - bakta have an option to list available database, so if a new database is present it could better to have the choice do download it, but how to deal with galaxy to give the more recent list of database ?
    - My first idea is to make a static list of tha last database version but if you have an idea to be alway update it could better.
 I don't understand how I can specify the output path of the download files ?

pimarin commented:
I try to make a good version for data_manager, but if the python code seem to work, I failed with the galaxy json output.

```
AssertionError: Invalid data manager (bakta_build_database) requested. It may have been removed before the job completed.
2022-10-14 11:55:17,897 
ERROR [galaxy.jobs] fail(): Missing output file in working directory: [Errno 2] No such file or directory: '/tmp/tmp0y80ol57/job_working_directory/000/1/outputs/galaxy_dataset_5850f29c-586d-475a-8b66-bff82d072fdc.dat'
```
But the file is present but not mooved in the output:
```
_outputs
│   └── galaxy_dataset_5850f29c-586d-475a-8b66-bff82d072fdc.dat
├── outputs
│   ├── COMMAND_VERSION
│   ├── tool_stderr
│   └── tool_stdout
```
And when I had a `cat command in the xml file to show the `'$outputfile'` the file content is okay

pimarin commented:
Also for the test part, I tried to mimic a link to a database test but I can't test all the code, for example the amrfinderplus tool method which call the amrfinder DB...

mvdbeek commented:
I'm not sure where you're seeing this, your current error is this:
```
Traceback (most recent call last):
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/jobs/__init__.py", line 1895, in finish
    self.app, inp_data, out_data, param_dict, job=job, final_job_state=final_job_state
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/tools/__init__.py", line 3026, in exec_after_process
    data_manager.process_result(out_data)
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/tools/data_manager/manager.py", line 351, in process_result
    data_table.add_entry(data_table_value, persist=True, entry_source=self)
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/tools/data/__init__.py", line 362, in add_entry
    self._add_entry(entry, allow_duplicates=allow_duplicates, persist=persist, entry_source=entry_source, **kwd)
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/tools/data/__init__.py", line 768, in _add_entry
    fields = self._replace_field_separators(fields)
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/tools/data/__init__.py", line 859, in _replace_field_separators
    return [x.replace(separator, replace) for x in fields]
  File "/tmp/tmp7os2nl9c/galaxy-dev/lib/galaxy/tools/data/__init__.py", line 859, in <listcomp>
    return [x.replace(separator, replace) for x in fields]
AttributeError: 'float' object has no attribute 'replace'
```
from https://github.com/galaxyproject/tools-iuc/actions/runs/3249976194/jobs/5333102888#step:8:693

pimarin commented:
I failed to mimic the same test env, for data_manager I'm using this line :
```
planemo test --biocontainers --no_dependency_resolution --no_conda_auto_init --galaxy_source https://github.com/galaxyproject/galaxy --galaxy_branch release_22.05 --galaxy_python_version 3.7
```

pimarin commented:
Okay I'm lost, I identify in the CI that file comparison failed due to the difference in the path, but I don't understand how I can remove it...

pimarin commented:
I don't understand, why the diff file give me an error:
```
Output output_file:  different than expected, difference (using diff):
( /tmp/tmpybqgde0gbakta_test_data_manager.json v. /tmp/tmp5gkyc2m0bakta_test_data_manager.json )
--- local_file
+++ history_data
@@ -1 +1 @@
-{"data_tables": {"bakta_database": [{"bakta_version": "0.0", "database_record": "7197299", "dbkey": "date_test", "path": "bakta_0.0", "value": "bakta_0.0"}]}}
+{"data_tables": {"bakta_database": [{"bakta_version": "0.0", "database_record": "7197299", "dbkey": "date_test", "path": "bakta_0.0", "value": "bakta_0.0"}]}}
```

bgruening commented:
@gregvonkuster can you please help here? Thanks a bunch!

gregvonkuster commented:
> @gregvonkuster can you please help here? Thanks a bunch!

@pimarin I'm coming in very late to this conversation, so what is it, in particular, that you need help with?  Is it the failing test immediately above?

pimarin commented:
Hi @gregvonkuster !
First, how can I use planemo test with your configuration  to test on my local machine ? I failed to have the same output as on github
Second, It seem to work but the diff file give me an error but the diff seem to be identical !
Finally, I d'ont undrstand how to deal with the output path to give in the data_manager, I write the `value`also in the `path` json file

gregvonkuster commented:
> Hi @gregvonkuster !
> First, how can I use planemo test with your configuration to test on my local machine ? I failed to have the same output as on github

Not sure what you mean about GitHub here, but to run planemo test on you local machine:
```
$cd <planemo install dir>
$. planemo/bin./activate
$cd <tool_dir?
$planemo test <my_tool>.xml
```
See the plane docs here https://planemo.readthedocs.io/en/latest/commands/test.html

> Second, It seem to work but the diff file give me an error but the diff seem to be identical !

So it sounds like you do have planemo test returning results.  So if the test results differ from your test output file, then there is something different about the 2 outputs.  It is sometimes difficult to visually see the differences due to whitespace, etc.

You can try the --no_cleanup flag so that planemo will keep the outputs.
```
$planemo test --no_cleanup <my_tool>.xml
```
You're see in th eplanemo log the temporary directory that planemo creates for running the tool and you can manually check the difference of the planemo output with your output.

> Finally, I d'ont undrstand how to deal with the output path to give in the data_manager, I write the `value`also in the `path` json file

You're discussing the test output json file, correct?  If so, the path in the test output json file is based upon the data that is being downloaded and how it is ultimately installed.  There are many data manager tools in this repo that you can take a look at to see how they work.

For example, if your URL downloads a set of files that should be placed in a directory, the path could be null in your output test json file.  Here is an example - https://github.com/galaxyproject/tools-iuc/blob/master/data_managers/data_manager_malt_index_builder/test-data/malt_index_builder1.json.  It's usually best to run your tool in a local Galaxy to see how it installs the data and the output JSON in the history item.  This will help you to figure out what the test output json file should look like.

Also, make sure you read the data manager tool docs here - https://docs.galaxyproject.org/en/latest/dev/data_managers.html

pimarin commented:
Dear @gregvonkuster , 
thank you to your comment, I'll read the doc for the data_manager.
But for planemo test, your continous integration has specific parameter related to the planemo test not just `planemo test` and i failed to have the same output between my computer and the PR on github.

pimarin commented:
I tried the `planemo test` on other data_manager such as pangolearn, but test failed with same erro as for bakta :
```
FAIL: ( bakta_build_database ) > Test-1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/tmpwv66s4og/galaxy-dev/test/functional/test_toolbox.py", line 115, in test_tool
    self.do_it(tool_version=tool_version, test_index=test_index)
  File "/tmp/tmpwv66s4og/galaxy-dev/test/functional/test_toolbox.py", line 49, in do_it
    register_job_data=register_job_data,
  File "/tmp/tmpwv66s4og/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1257, in verify_tool
    raise e
  File "/tmp/tmpwv66s4og/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1252, in verify_tool
    testdef, test_history, jobs, data_list, data_collection_list, galaxy_interactor, quiet=quiet
  File "/tmp/tmpwv66s4og/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1449, in _verify_outputs
    raise JobOutputsError(found_exceptions, job_stdio)
galaxy.tool_util.verify.interactor.JobOutputsError: Job in error state.. tool_id: bakta_build_database, exit_code: 127, stderr: /bin/bash: /tmp/tmpwv66s4og/job_working_directory/000/1/tool_script.sh: No such file or directory
.
Job in error state.. tool_id: bakta_build_database, exit_code: 127, stderr: /bin/bash: /tmp/tmpwv66s4og/job_working_directory/000/1/tool_script.sh: No such file or directory
.
-------------------- >> begin captured stdout << ---------------------
Problem in history with id c47e0e7d8f946370 - summary of history's datasets and jobs below.

```

pimarin commented:
I used a symlonk to the json output. Even if planemo tests failed all the time, if I install the data_manager from the testtooplshed on a local galaxy (22.05), the tool seem to work, it download install and I can use the database in the bakta tool...

wm75 commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.py:
```diff
@@ -0,0 +1,258 @@
+                  f"calculated md5 = {md5.hexdigest()}"
+                  f" different from {self.md5} ")
+
+    def update_amrfinderplus_db(self):
+        amrfinderplus_db_path = f"{self.db_dir}/{self.db_name}/db/amrfinderplus-db"
+        if self.db_version == "test":
+            cmd = [
+                'amrfinder_update',
+                '--database', str(amrfinderplus_db_path),
+                '--force_update',
+                '--help'
+            ]
+        else:
+            cmd = [
+                'amrfinder_update',
+                '--database', str(amrfinderplus_db_path),
+                '--force_update'
+            ]
+        proc = sp.run(
+            cmd,
+            universal_newlines=True
+        )
+        if proc.returncode != 0:
+            print(f"ERROR: AMRFinderPlus failed! "
+                  f"command: 'amrfinder_update --force_update"
+                  f" --database {amrfinderplus_db_path}'")
+        else:
+            print("AMRFinderPlus database download")
```
This is *very* questionable. One of the main points of data managers is reproducibility, but doing this on the fly update of the amrfinder DB means that the "same" installed bakta DB on different instances of Galaxy could actually differ.

My intuition would be to leave the amrfinder DB untouched here and, instead, write a separate amrfinder data manager that the admin could use to install newer versions of that DB. In the bakta *tool* then you could have an option to use a specific amrfinder DB from that other data table instead of the one that comes with the bakta DB.

wm75 replied:
A requirement for doing this is that bakta, the tool, needs to be able to work with a symlinked DB (or one in which certain parts are replaced with symlinks) since the contents would then come from two different locations on the Galaxy server.

pimarin replied:
If I understood, 
Make a DM for bakta DB only
Make another DM for AMRfinderplus
and add in the tool, the option to choose both bakta and armfinderpluys version of the DB ?

wm75 replied:
Yes, and make the armfinderplus selection optional with the default being to use the one coming with the bakta download.

pimarin replied:
Okay I'll try, amrfinder_plus data manager already exist but not in the iuc repository, but still available on the european tool-shed for example. Is it possible to link with my tool ? [amrfinderplus](https://toolshed.g2.bx.psu.edu/view/estrain/data_manager_amrfinderplus/1ce80c2938fe)

pimarin replied:
bakta tool have ab option to specify path, I can make symlink into the wrapper to point bakta + amrfinder on the same location to the tool

wm75 commented on data_managers/data_manager_build_bakta_database/test-data/bakta_test.loc:
```diff
@@ -0,0 +1,7 @@
+# value, dbkey, database_record, bakta_version, path
+#
+# for example
+bakta_0.0    date_test    7197299    0.0    ${__HERE__}
```
The separators here need to be "real" tabs. You've got spaces!

wm75 replied:
This was resolved but somehow one of your commits since then must have reverted back to spaces!

wm75 commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.xml:
```diff
@@ -0,0 +1,45 @@
+    <expand macro="version_command"/>
+
+    <command detect_errors="aggressive"><![CDATA[
+        mkdir './output_dir_json' &&
+        ln -s '$output_file' './output_dir_json/json_output.json' &&
+        python '$__tool_directory__/bakta_build_database.py'
+        './output_dir_json/json_output.json'
```
This symlinking business should not be required, but I haven't yet fully understood all the issues you've encountered.

wm75 reviewed (commented):
Here's an idea for handling the test case more elegantly:

- introduce a parameter of `type="hidden"` with a `value=""` 
- in the test, set this invisible param to "--test"
- in the command section pass the param as a flag to the python script
- in `--test` mode run with a hard-coded json like now, but with a completely unrelated and small .tar.gz from somewhere (and its correct md5 sum)
  remember: the goal here is not to test bakta, but only that the logic of the data manager runs without an issue and produces the expected json, so the contents of the tar archive don't matter (and there's no point in downloading > 30 GB just for CI)

In addition the tool needs to skip the amrfinder DB update in `--test` mode (but I'm arguing that shouldn't be done in the first place anyway)

wm75 commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.py:
```diff
@@ -0,0 +1,258 @@
+
+    def get_data_manager(self, bakta_database_info, output_path):
+        self.bakta_table_list = self.get_data_table_format()
+        bakta_value = f"bakta_{bakta_database_info['major']}." \
+                      f"{bakta_database_info['minor']}"
+        data_info = dict(value=bakta_value,
+                         dbkey=bakta_database_info['date'],
+                         database_record=bakta_database_info['record'],
```
Storing both your composed bakta_value and the database record field seems redundant. Wouldn't the database record serve as an excellent value?

pimarin replied:
You're right, but using major and minor value is more explicit to user no ?
And I remove the record ?

wm75 replied:
Hmm, good point. So there are two different aspects here:
1. Each record in a data table needs a unique ID (stored in the value column)
2. Each record should have a nice representation to display to the user in a select menu

I agree that the major.minor version is what you'd want for point number 2 and you can store this value in a "name" or "db_version" or whatever column.
For the "value" column I would use the 'record' ID cause its unique by nature, and, yes, you can drop the record column then.

=> You won't save a column as I thought originally, but you'd move record to be the value and store the DB version in an extra column instead.

pimarin commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.py:
```diff
@@ -0,0 +1,258 @@
+                for db_dic in versions:
+                    db_date_list.append(datetime.strptime(db_dic["date"],
+                                                          '%Y-%m-%d').date())
+                max(db_date_list)
```
I @wm75 , thank you to your help !
The fetch_db_versions make a list of available version but for the latest I used the most recent database in the liste

wm75 replied:
Yes, I know. Just wanted to point out that the expression `max(db_date_list)` doesn't do anything useful on its own here.

pimarin replied:
you right ! I d'ont now why I forgot to remove it !

pimarin commented:
with this :
` planemo test --biocontainers --no_dependency_resolution --no_conda_auto_init  --galaxy_source https://github.com/galaxyproject/galaxy --galaxy_branch release_22.05 --galaxy_python_version 3.7
`
I obtain
```
Problems:

    Job in error state.. tool_id: bakta_build_database, exit_code: 127, stderr: /bin/bash: /tmp/tmpu0o448l2/job_working_directory/000/1/tool_script.sh: No such file or directory
    .

    Job in error state.. tool_id: bakta_build_database, exit_code: 127, stderr: /bin/bash: /tmp/tmpu0o448l2/job_working_directory/000/1/tool_script.sh: No such file or directory
    .

Command Line:

    python '/home/pierre/.planemo/planemo_tmp_qnh7zu_3/data_manager/bakta_build_database.py' '/tmp/tmpu0o448l2/job_working_directory/000/1/outputs/galaxy_dataset_73ef38f4-c008-4802-b32b-937acd668eb6.dat' --database_version 'latest' '--test'

Exit Code:

    127

Standard Error:

    /bin/bash: /tmp/tmpu0o448l2/job_working_directory/000/1/tool_script.sh: No such file or directory
```

pimarin commented:
```
python '/home/pierre/.planemo/planemo_tmp_qnh7zu_3/data_manager/bakta_build_database.py' '/tmp/tmpu0o448l2/job_working_directory/000/1/outputs/galaxy_dataset_73ef38f4-c008-4802-b32b-937acd668eb6.dat' --database_version 'latest' '--test']
2022-11-02 11:43:58,696 INFO  [galaxy.tool_util.output_checker] Job error detected, failing job. Reasons are [{'type': 'exit_code', 'desc': 'Fatal error: Exit code 127 ()', 'exit_code': 127, 'code_desc': '', 'error_level': 3}]
/tmp/tmpu0o448l2/galaxy-dev/lib/galaxy/model/__init__.py:454: SAWarning: Dialect sqlite+pysqlite does *not* support Decimal objects natively, and SQLAlchemy must convert from floating point - rounding errors and other issues may occur. Please consider storing Decimal numbers as strings or integers on this platform for lossless storage.
  return self.text_metrics + self.numeric_metrics
FAIL

======================================================================
FAIL: ( bakta_build_database ) > Test-1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/tmpu0o448l2/galaxy-dev/test/functional/test_toolbox.py", line 115, in test_tool
    self.do_it(tool_version=tool_version, test_index=test_index)
  File "/tmp/tmpu0o448l2/galaxy-dev/test/functional/test_toolbox.py", line 49, in do_it
    register_job_data=register_job_data,
  File "/tmp/tmpu0o448l2/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1257, in verify_tool
    raise e
  File "/tmp/tmpu0o448l2/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1252, in verify_tool
    testdef, test_history, jobs, data_list, data_collection_list, galaxy_interactor, quiet=quiet
  File "/tmp/tmpu0o448l2/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1449, in _verify_outputs
    raise JobOutputsError(found_exceptions, job_stdio)
galaxy.tool_util.verify.interactor.JobOutputsError: Job in error state.. tool_id: bakta_build_database, exit_code: 127, stderr: /bin/bash: /tmp/tmpu0o448l2/job_working_directory/000/1/tool_script.sh: No such file or directory
.
Job in error state.. tool_id: bakta_build_database, exit_code: 127, stderr: /bin/bash: /tmp/tmpu0o448l2/job_working_directory/000/1/tool_script.sh: No such file or directory
.
```

wm75 commented:
So the main issue that the test failures are pointing to is this:
https://github.com/galaxyproject/tools-iuc/blob/b082e89a0090f9c4b2a257bb06b7c6f141b42650/data_managers/data_manager_build_bakta_database/data_manager_conf.xml#L12
The path you're recording and returning in the json file should not be absolute. It's up to Galaxy to figure out where to put the managed data and it will do so with the help of the data_manager_conf.xml. This way then the content of the json file should become fully predictable.

wm75 commented:
The relevant docs for the important section of the data_manager_conf.xml is https://galaxyproject.org/admin/tools/data-managers/data-manager-xml-syntax/#move-tag-set and below.

wm75 commented:
In brief, the <move> tag describes where to move downloaded data and <value_translation> describes how the json's path should be recorded in the final data table.

pimarin commented:
I failed to understand how to fix the problem for output...

- In the data_manager.xml I specified a `type=directory`
- I added the source with `${path}` which is the path from the json file and this path is just `"."`
- `relativize_symlinks="True"` I don't understand this option
```
                <column name="path" output_ref="output_file">
                    <move type="directory" relativize_symlinks="True">
                        <source>${path}</source>
                        <target base="${GALAXY_DATA_MANAGER_DATA_PATH}">bakta_database/${value}</target>
                    </move>
                    <value_translation>${GALAXY_DATA_MANAGER_DATA_PATH}/bakta_database/${value}</value_translation>
                    <value_translation type="function">abspath</value_translation>
                </column>
```
- I extracted from the json file the path 
```
    with open(all_args.data_manager_json) as fh:
        params = json.load(fh)
    target_dir = params['output_data'][0]['extra_files_path']
```
The path returned is like : /tmp/path/to/dataset_009dd27e-b2a3-4636-ad16-f8ca4ff967e0_files`
The database is downloaded into with success, untar but after nothing happened.
```
├── _configs
├── _outputs
│   └── galaxy_dataset_009dd27e-b2a3-4636-ad16-f8ca4ff967e0.dat
├── outputs
│   ├── tool_stderr
│   └── tool_stdout
├── tool_script.sh
├── _working
└── working
    ├── conda_activate.log
    └── dataset_009dd27e-b2a3-4636-ad16-f8ca4ff967e0_files
        ├── db
        │   └── version.txt
        └── db.tar.gz
```
I don't understand the error I made between data_manager and python script

pimarin commented:
And when I try to use planemo test on other data_manager in the IUC repository it failed also

pimarin commented:
for example I started `planemo t`with metaphlan DM :
```
} ; python '/home/pierre/.planemo/planemo_tmp_k5yckajt/data_manager/data_manager_metaphlan_download.py' --index 'mpa_v30_CHOCOPhlAn_201901' --json '/tmp/tmpoqcnh54x/job_working_directory/000/1/outputs/galaxy_dataset_377bf178-b80d-47c4-8750-1ac0fe69b4fb.dat']
2022-11-03 14:52:35,079 ERROR [galaxy.jobs] exec_after_process hook failed for job 1
Traceback (most recent call last):
  File "/tmp/tmpoqcnh54x/galaxy-dev/lib/galaxy/jobs/__init__.py", line 1895, in finish
    self.app, inp_data, out_data, param_dict, job=job, final_job_state=final_job_state
  File "/tmp/tmpoqcnh54x/galaxy-dev/lib/galaxy/tools/__init__.py", line 3025, in exec_after_process
    ), f"Invalid data manager ({data_manager_id}) requested. It may have been removed before the job completed."
AssertionError: Invalid data manager (data_manager_metaphlan_download) requested. It may have been removed before the job completed.
2022-11-03 14:52:35,085 ERROR [galaxy.jobs] fail(): Missing output file in working directory: [Errno 2] No such file or directory: '/tmp/tmpoqcnh54x/job_working_directory/000/1/outputs/galaxy_dataset_377bf178-b80d-47c4-8750-1ac0fe69b4fb.dat'
/tmp/tmpoqcnh54x/galaxy-dev/lib/galaxy/model/__init__.py:454: SAWarning: Dialect sqlite+pysqlite does *not* support Decimal objects natively, and SQLAlchemy must convert from floating point - rounding errors and other issues may occur. Please consider storing Decimal numbers as strings or integers on this platform for lossless storage.
  return self.text_metrics + self.numeric_metrics
FAIL

======================================================================
FAIL: ( data_manager_metaphlan_download ) > Test-1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/tmpoqcnh54x/galaxy-dev/test/functional/test_toolbox.py", line 115, in test_tool
    self.do_it(tool_version=tool_version, test_index=test_index)
  File "/tmp/tmpoqcnh54x/galaxy-dev/test/functional/test_toolbox.py", line 49, in do_it
    register_job_data=register_job_data,
  File "/tmp/tmpoqcnh54x/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1257, in verify_tool
    raise e
  File "/tmp/tmpoqcnh54x/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1252, in verify_tool
    testdef, test_history, jobs, data_list, data_collection_list, galaxy_interactor, quiet=quiet
  File "/tmp/tmpoqcnh54x/galaxy-dev/lib/galaxy/tool_util/verify/interactor.py", line 1449, in _verify_outputs
    raise JobOutputsError(found_exceptions, job_stdio)
galaxy.tool_util.verify.interactor.JobOutputsError: Job in error state.. tool_id: data_manager_metaphlan_download, exit_code: 0, stderr: 
Downloading MetaPhlAn database
Please note due to the size this might take a few minutes
```

wm75 commented:
Don't change too many things at the same time.
[44229e7](https://github.com/galaxyproject/tools-iuc/pull/4723/commits/44229e79b3b453539e81ac1c71223c1c40e59c6a) was very close! The error message there was
```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.15/x64/lib/python3.7/shutil.py", line 566, in move
    os.rename(src, real_dst)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpc5ul0_e4/files/000/dataset_1_files/.' -> '/tmp/tmpc5ul0_e4/galaxy-dev/tool-data/bakta_database/7197299'
```
which makes some sense. So if you go from there and **only** change the path from '.' to 'db' **and** remove the non-helpful `compare="contains"` from the test, you'll start seeing a useful diff from which you can start working.

wm75 commented on data_managers/data_manager_build_bakta_database/data_manager/bakta_build_database.py:
```diff
@@ -0,0 +1,255 @@
+                         dbkey=bakta_value,
+                         bakta_version=tool_version,
+                         path="db")
+        self.bakta_table_list["data_tables"][self.data_table_name] = data_info
```
```suggestion
        self.bakta_table_list["data_tables"][self.data_table_name] = [data_info]
```

A DM could also add several records to a data table in a single run so this needs to be a list!

pimarin replied:
Thank you @wm75 !
Building a DM seem to not totally clear to me, I hope improve it for the next DM.
But I don't understand how to reproduce the same result for planemo test on my local machine,
When I use :
`planemo test --biocontainers --no_dependency_resolution --no_conda_auto_init --galaxy_source https://github.com/galaxyproject/galaxy --galaxy_branch release_22.05 --galaxy_python_version 3.7 --no_cleanup`
which seem to be very close to the test used on github in CI, I can't have the same output

pimarin commented:
Now I have just a diff error but I can't use the `--update_test_data` on my local PC because I failed to have the same test as on github

wm75 commented:
Ok, I think I solved the mystery about the diff showing seemingly identical lines:
```
--- a/data_managers/data_manager_build_bakta_database/test-data/bakta_test_data_manager.json
+++ b/data_managers/data_manager_build_bakta_database/test-data/bakta_test_data_manager.json
@@ -1 +1 @@
-{"data_tables": {"bakta_database": [{"bakta_version": "0.0", "dbkey": "V0.0_date_test", "path": "db", "value": "7197299"}]}}
+{"data_tables": {"bakta_database": [{"bakta_version": "0.0", "dbkey": "V0.0_date_test", "path": "db", "value": "7197299"}]}}
\ No newline at end of file
```

It's a bit silly but the json generated by the data manager doesn't have a newline at the end of the file and that does in fact matter for the test. Standard text editors (with the exception of emacs I think) will, however always terminate files with \n so if you ever save the test file from such an editor, future tests will fail.

wm75 commented:
I've fixed the file for you with planemo update_test_data because I have no difficulties obtaining the same output as the CI here. Just remember to pull again before continuing to work on this.

wm75 commented:
- https://github.com/galaxyproject/tools-iuc/pull/4723#discussion_r1008172370
- https://github.com/galaxyproject/tools-iuc/pull/4723#discussion_r1008174200
- https://github.com/galaxyproject/tools-iuc/pull/4723#discussion_r1008197718
- https://github.com/galaxyproject/tools-iuc/pull/4723#discussion_r1008199900
- https://github.com/galaxyproject/tools-iuc/pull/4723#discussion_r1008200390
-https://github.com/galaxyproject/tools-iuc/pull/4723#discussion_r1014498750

need to be resolved still. After that I can do a second pass reviewing the basically working and cleaned up code :)

pimarin commented:
Okay cool ! this work (only on github, on my local test still failed)

pimarin commented:
I have another question, 
Bakta database availability depends on the bakta tool version, (eg.  bakta 1.5.1 accept only the 4.0 version).
I made a filter in the bakta tool to see only the available database version, but I wondering if the filter must be add in the data manager

bernt-matthias commented:
> wondering if the filter must be add in the data manager

I think the DM only needs to make sure that the correct database version is stored in the data table.

pimarin commented:
The DM is okay ?

pimarin commented:
> > wondering if the filter must be add in the data manager
> 
> I think the DM only needs to make sure that the correct database version is stored in the data table.

With another test on another database_test version ?

bernt-matthias commented on data_managers/data_manager_build_bakta_database/test-data/bakta_test_data_manager.json:
```diff
@@ -0,0 +1 @@
+{"data_tables": {"bakta_database": [{"bakta_version": "0.0", "dbkey": "V0.0_date_test", "path": "db", "value": "7197299"}]}}
```
Wondering what `"value": "7197299"` means? Shouldn't this be a path?

pimarin replied:
this is a unique value provided from the zenodo website, it could be added in the path to be also unique

pimarin replied:
this is the part of the DOI

bernt-matthias commented:
You will need to adapt the data table of the bakta tool: 

https://github.com/galaxyproject/tools-iuc/blob/0ef6e6b2dcc9a41b32741368a60b931f31934686/tools/bakta/tool_data_table_conf.xml.sample

Also the bakta tool paramater could need a validator checking the version in the data table.
